# Nivell 1

## Pregunta 1

Executa el següent codi en una cel·la:

```java
short s = (short)32000;
short s = (short)35000;
```

Quins resultats obtens? Per què?

In [7]:
short s = (short)32000;
short s = (short)35000;

System.out.print(s);

-30536

Resposta:

El resultado que da es "-30536" porque el tipo de dato que utilizamos es un "short" que tiene un límite de rango (-32 768 a 32 767), entonces el primero al hacer System.out.print(), sacaria el resultado igual que esta porque esta dentro de los limites permitodos del "short", pero el siguiente está fuera de rango de lo permitido, entonces lo que pasando es que la el último valor permitido postivo que es el 32767 es el máximo, no va seguir 32768, 32679, 32770... sino que va empecar de nuevo desde el valor rango negativo que es el -32768, entonces sale el resultado "-30536" porque se le sumo al valor "-32768" que es el valor minimo del rango el valor de la resta que esta fuera de rango, es decir: 35000-32768 = 2232 -> y ahora se le suma al valor minimo de rango el resultado de resta: -32768 + 2232 = "-30536"

## Pregunta 2

Executa el següent codi en una cel·la:

```java
var v = 5.0;
var v0 = 5;
```

De quin tipus de dada són les variables v i v0?

In [14]:
var v = 5.0;
var v0 = 5;

System.out.println(v);
System.out.print(v0);

5.0
5

Resposta:

La variable "v": es un "double"       
La variable "v0": es un "int"

Ara, executa:

```java
var v1 = v + v0;
```
De quin tipus de dada és la variable v1? Per què (què està passant en fer-se la suma)?

In [16]:
var v1 = v + v0;
System.out.print(v1);

10.0

Resposta:

La variable "v1": es de tipo "double"

Porqué hemos sumado un "double" y un "int" promocionado a double, como resultado se detectó como un valor decimal, que lo identifica como double.

## Exercici 1

Escriu una funció, la signatura de la qual serà

```java
void isPrime(int n);
```
que imprimeixi ```true``` si n és un número primer positiu, o ```false``` si n és un número no primer o bé és 0, o bé és un enter negatiu.

Per exemple:

```java
isPrime(7); //Imprimirà true
isPrime(20); //Imprimirà false
isPrime(0); //Imprimirà false
isPrime(-7); //Imprimirà false
```

In [142]:
void isPrime(int n) {
    boolean esPrimo = true;
    if (n <= 1) {
        esPrimo = false; // si son negativos, directamente se consideran que no son numeros primos
    } else {
        for (int i = 2; i < n; i++) {
            if (n % i == 0) {
                esPrimo = false;
                break;
            }
        }
    }
    System.out.println(esPrimo);
}

// Pruebas aquí
isPrime(7); 
isPrime(20); 
isPrime(0);
isPrime(-7);
System.out.println(); 
// Mas pruebas
isPrime(3);
isPrime(4);
isPrime(33);
isPrime(11);
isPrime(24);

true


false
false
false

true
false
false
true
false


## Exercici 2

Escriu una funció, la signatura de la qual serà

```java
int getSum(int[] n);
```

que retorni un enter resultat de la suma de tots els enters que composen l'array.

In [154]:
//Escriu i prova el teu codi aquí

int getSum(int[] n) {
    int sumaTotal = 0;
    for(int i = 0; i <= n.length-1; i++) {
        sumaTotal+=n[i];
    }
    return sumaTotal;
}

// JOC DE PROVES
int[] array1 = {2, 33, 1, 14, 56, 102};
int[] array2 = {5, 28, 7, 92, 10, 50};
int[] array3 = {17, 3, 81, 9, 27, 42};
int[] array4 = {63, 37, 6, 12, 77, 18};
int[] array5 = {19, 4, 88, 25, 48, 70};

System.out.println(getSum(array1));
System.out.println(getSum(array2));
System.out.println(getSum(array3));
System.out.println(getSum(array4));
System.out.println(getSum(array5));


208
192
179
213
254


## Exercici 3

Tenim la següent classe:

```java
class Propulsor{
    private int potencia; //potencia actual que té en cada moment
    
    //potència màxima que pot tenir el propulsor
    //És final perquè un cop creat l'objecte, no es pot modificar
    private final int potenciaMaxima; 
    
    public Propulsor(int potenciaMaxima){
        this.potenciaMaxima = potenciaMaxima;
    }
    
    //setter
    //(potenciaMaxima no té setter perquè és final)
    public void setPotencia(int potencia){
        this.potencia = potencia;
    }
    
    //getters
    public int getPotencia(){
        return this.potencia;
    }
    
    public int getPotenciaMaxima(){
        return this.potenciaMaxima;
    }
}
```

Crea una classe ```Coet``` que tingui dos atributs: un identificador anomenat ```id``` de tipus String, i un propulsor anomenar ```propulsor``` de tipus Propulsor.

- La classe coet ha de tenir un constructor que inicialitzi l'objecte passant-li un id i un propulsor.
- L'id serà immutable (no es pot canviar un cop creat el coet), però el propulsor sí es podrà canviar (amb un setter). Amndós seràn privats, però s'han de poder llegir amb els getters.

**Has d'afegir, al coet, dos mètodes públics** (```arrenca``` i ```atura```), sense paràmetres.

> Proves: Crearàs un coet amb id “32WESSDS” que tingui un propulsor amb potència màxima 50.

- ```arrenca()``` no retorna res i ha d'imprimir per pantalla "Coet amb id 32WESSDS en marxa amb un propulsor de potència màxima 50".
- ```atura()``` no retorna res i ha d'imprimir per pantalla "Coet amb id 32WESSDS està aturat".

In [21]:
class Propulsor{
    private int potencia; 
    private final int potenciaMaxima; 
    
    public Propulsor(int potenciaMaxima){
        this.potenciaMaxima = potenciaMaxima;
    }
    
    // setter
    public void setPotencia(int potencia){
        this.potencia = potencia;
    }
    
    //getters
    public int getPotencia(){
        return this.potencia;
    }
    
    public int getPotenciaMaxima(){
        return this.potenciaMaxima;
    }
}

class Coet {
    private String id;
    private Propulsor propulsor;

    public Coet(String id, Propulsor propulsor) {
        this.id = id;
        this.propulsor = propulsor;
    }

    // setter
    public void setId(String id) {
        this.id = id;
    }

    public void setPropulsor(Propulsor propulsor) {
        this.propulsor = propulsor;
    }

    // getter
    public String getId() {
        return this.id;
    }

    public Propulsor getPropulsor() {
        return this.propulsor;
    }

    public void arrenca() {
        System.out.println("Coet amb id " + id + " en marxa amb un propulsor de potència màxima " + propulsor.getPotenciaMaxima());
    }

    public void atura() {
        System.out.println("Coet amb id " + id + " està aturat");
    }
}
String id = "32WESSDS";

Coet coete = new Coet(id, new Propulsor(50));
coete.arrenca();
coete.atura();

Coet amb id 32WESSDS en marxa amb un propulsor de potència màxima 50
Coet amb id 32WESSDS està aturat
